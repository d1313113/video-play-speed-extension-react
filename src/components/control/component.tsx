import React, { FC } from "react";
import { Slider, Button } from "antd";
import {
  PlayCircleOutlined,
  PauseCircleOutlined,
  AudioMutedOutlined,
  AudioOutlined
} from "@ant-design/icons";
import { VideoStatus } from "@/content-script/actions/videosPlaybackRate";
import "./styles.scss";

const RATES = [1, 2, 3, 4, 5];

export interface ControlProps {
  videoList: VideoStatus[];
  onSetRate: (idx: number, value: number) => Promise<void>;
  onPlay: (idx: number) => Promise<void>;
  onPause: (idx: number) => Promise<void>;
  onMuted: (idx: number, muted: boolean) => Promise<void>;
}

export const Control: FC<ControlProps> = (props: ControlProps) => {
  const handleSetVideoRate = (index: number, value: number): void => {
    props.onSetRate(index, value);
  };

  const handleClickRate = (index: number, rate: number): void => {
    props.onSetRate(index, rate);
  };

  const { videoList } = props;

  return (
    <main className="control">
      <div className="title">可拖动step调整速度</div>
      <section className="control-main">
        {videoList.map((item, index) => (
          <div className="video-item" key={index}>
            <Slider
              onChange={(value: number): void =>
                handleSetVideoRate(index, value)
              }
              min={0.1}
              max={5}
              step={0.1}
              defaultValue={1}
              value={item.playbackRate}
            />
            <div className="rate-wrapper">
              {item.paused ? (
                <PlayCircleOutlined
                  className="icon-control"
                  onClick={(): Promise<void> => props.onPlay(index)}
                />
              ) : (
                <PauseCircleOutlined
                  className="icon-control"
                  onClick={(): Promise<void> => props.onPause(index)}
                />
              )}
              {item.muted ? (
                <AudioMutedOutlined
                  className="icon-control middle"
                  onClick={(): Promise<void> =>
                    props.onMuted(index, !item.muted)
                  }
                />
              ) : (
                <AudioOutlined
                  className="icon-control middle"
                  onClick={(): Promise<void> =>
                    props.onMuted(index, !item.muted)
                  }
                />
              )}
              {RATES.map(rate => (
                <Button
                  key={rate}
                  onClick={(): void => handleClickRate(index, rate)}
                  size="small"
                  type={rate === 1 ? "primary" : "default"}
                  disabled={rate === item.playbackRate}
                >
                  {rate === 1 ? "Reset" : `✖️${rate}`}
                </Button>
              ))}
            </div>
          </div>
        ))}
      </section>
    </main>
  );
};
