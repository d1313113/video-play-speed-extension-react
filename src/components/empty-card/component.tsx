import React, { FC } from "react";
import { Empty } from "antd";
import "./styles.scss";

export const EmptyCard: FC = () => (
  <section className="empty-wrapper">
    <Empty className="empty" description="当前页面未找到视频:)" />
  </section>
);
