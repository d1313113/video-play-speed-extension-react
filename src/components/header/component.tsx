import React, { FC, useState } from "react";
import { ReloadOutlined } from "@ant-design/icons";
import "./styles.scss";

interface HeaderProps {
  onReload: () => void;
}

export const Header: FC<HeaderProps> = (props: HeaderProps) => {
  const [spin, setSpin] = useState<boolean>(false);

  const handleReload = (): void => {
    setSpin(!spin);
    if (!spin) {
      props.onReload();
      setTimeout((): void => {
        setSpin(false);
      }, 1000);
    }
  };

  return (
    <header className="header-wrapper">
      <div className="header-content">
        <div className="title">视频播放速度调节插件</div>
        <div className="sub-title">可以让你轻松调整视频播放速度</div>
      </div>
      <ReloadOutlined
        onClick={handleReload}
        className="icon-reload"
        spin={spin}
        style={{
          fontSize: "20px"
        }}
      />
    </header>
  );
};
