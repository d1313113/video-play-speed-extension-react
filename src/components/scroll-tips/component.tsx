import React, { FC } from "react";
import { DoubleRightOutlined } from "@ant-design/icons";
import "./styles.scss";

export const ScrollTips: FC = () => (
  <section className="tips-wrapper">
    <div className="tips">Scroll to load more</div>
    <DoubleRightOutlined
      className="icon-tips"
      rotate={90}
      style={{ fontSize: "20px" }}
    />
  </section>
);
