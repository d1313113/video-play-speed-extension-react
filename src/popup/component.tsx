import React, { FC, useEffect, useState } from "react";
import { browser } from "webextension-polyfill-ts";
import { throttle } from "lodash";
import getCurrentTab from "@/utils/getCurrentTab";
import {
  actionTypes,
  VideoStatus,
  $$
} from "@/content-script/actions/videosPlaybackRate";

import { Header } from "@/components/header";
import { EmptyCard } from "@/components/empty-card";
import { Control } from "@/components/control";
import { ScrollTips } from "@/components/scroll-tips";
import "./styles.scss";

export const Popup: FC = () => {
  const [videoList, setVideoList] = useState<VideoStatus[]>([]);
  const [scrollTop, setScrollTop] = useState<number>(0);

  const initPopUp = async (): Promise<void> => {
    console.log(11111);
    const { id: tabId } = await getCurrentTab();
    if (tabId != null) {
      const resp: VideoStatus[] = await browser.tabs.sendMessage(tabId, {
        action: actionTypes.GET_VIDEOS_STATUS
      });
      setVideoList(resp);
    }
  };

  const handleSetVideoRate = async (
    idx: number,
    value: number
  ): Promise<void> => {
    const { id: tabId } = await getCurrentTab();

    if (tabId != null) {
      await browser.tabs.sendMessage(tabId, {
        action: actionTypes.SET_VIDEO_PLAYBACK_RATE,
        idx,
        value
      });
      initPopUp();
    }
  };

  const handleSetVideoPlay = async (idx: number): Promise<void> => {
    const { id: tabId } = await getCurrentTab();

    if (tabId != null) {
      await browser.tabs.sendMessage(tabId, {
        action: actionTypes.SET_VIDEO_PLAY,
        idx
      });
      initPopUp();
    }
  };

  const handleSetVideoPause = async (idx: number): Promise<void> => {
    const { id: tabId } = await getCurrentTab();

    if (tabId != null) {
      await browser.tabs.sendMessage(tabId, {
        action: actionTypes.SET_VIDEO_PAUSE,
        idx
      });
      initPopUp();
    }
  };

  const handleScroll = (html: HTMLElement): void => {
    try {
      const { scrollTop } = html;
      setScrollTop(scrollTop);
    } catch (e) {
      console.error(e);
    }
  };

  // Sends the `popupMounted` event
  useEffect(() => {
    browser.runtime.sendMessage({ popupMounted: true });
    initPopUp();
    const html = $$<HTMLElement>("html")[0];
    const _throttle = throttle((): void => handleScroll(html), 100);
    window.addEventListener("scroll", _throttle, false);

    return (): void => {
      window.removeEventListener("scroll", _throttle, false);
    };
  }, []);

  // Renders the component tree
  return (
    <div className="popup-container">
      <div className="container">
        <Header onReload={initPopUp} />
        {!videoList.length ? (
          <EmptyCard />
        ) : (
          <Control
            videoList={videoList}
            onSetRate={handleSetVideoRate}
            onPlay={handleSetVideoPlay}
            onPause={handleSetVideoPause}
          />
        )}
        {!!videoList.length && videoList.length > 5 && scrollTop < 30 && (
          <ScrollTips />
        )}
      </div>
    </div>
  );
};
