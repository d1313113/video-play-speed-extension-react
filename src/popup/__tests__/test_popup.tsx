import * as React from "react";
import { Popup } from "../component";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";

it("component renders", () => {
  const tree:
    | ReactTestRendererJSON
    | ReactTestRendererJSON[]
    | null = renderer.create(<Popup />).toJSON();
  expect(tree).toMatchSnapshot();
});
