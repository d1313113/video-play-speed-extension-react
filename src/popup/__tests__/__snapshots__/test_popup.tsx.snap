// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders 1`] = `
<div
  className="popup-container"
>
  <div
    className="container"
  >
    <header
      className="header-wrapper"
    >
      <div
        className="header-content"
      >
        <div
          className="title"
        >
          视频播放速度调节插件
        </div>
        <div
          className="sub-title"
        >
          可以让你轻松调整视频播放速度
        </div>
      </div>
      <span
        aria-label="reload"
        className="anticon anticon-reload icon-reload"
        onClick={[Function]}
        role="img"
        style={
          Object {
            "fontSize": "20px",
          }
        }
        tabIndex={-1}
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="reload"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M909.1 209.3l-56.4 44.1C775.8 155.1 656.2 92 521.9 92 290 92 102.3 279.5 102 511.5 101.7 743.7 289.8 932 521.9 932c181.3 0 335.8-115 394.6-276.1 1.5-4.2-.7-8.9-4.9-10.3l-56.7-19.5a8 8 0 00-10.1 4.8c-1.8 5-3.8 10-5.9 14.9-17.3 41-42.1 77.8-73.7 109.4A344.77 344.77 0 01655.9 829c-42.3 17.9-87.4 27-133.8 27-46.5 0-91.5-9.1-133.8-27A341.5 341.5 0 01279 755.2a342.16 342.16 0 01-73.7-109.4c-17.9-42.4-27-87.4-27-133.9s9.1-91.5 27-133.9c17.3-41 42.1-77.8 73.7-109.4 31.6-31.6 68.4-56.4 109.3-73.8 42.3-17.9 87.4-27 133.8-27 46.5 0 91.5 9.1 133.8 27a341.5 341.5 0 01109.3 73.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 003 14.1l175.6 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c-.1-6.6-7.8-10.3-13-6.2z"
          />
        </svg>
      </span>
    </header>
    <section
      className="empty-wrapper"
    >
      <div
        className="ant-empty empty"
      >
        <div
          className="ant-empty-image"
        >
          <svg
            className="ant-empty-img-default"
            height="152"
            viewBox="0 0 184 152"
            width="184"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fillRule="evenodd"
            >
              <g
                transform="translate(24 31.67)"
              >
                <ellipse
                  className="ant-empty-img-default-ellipse"
                  cx="67.797"
                  cy="106.89"
                  rx="67.797"
                  ry="12.668"
                />
                <path
                  className="ant-empty-img-default-path-1"
                  d="M122.034 69.674L98.109 40.229c-1.148-1.386-2.826-2.225-4.593-2.225h-51.44c-1.766 0-3.444.839-4.592 2.225L13.56 69.674v15.383h108.475V69.674z"
                />
                <path
                  className="ant-empty-img-default-path-2"
                  d="M101.537 86.214L80.63 61.102c-1.001-1.207-2.507-1.867-4.048-1.867H31.724c-1.54 0-3.047.66-4.048 1.867L6.769 86.214v13.792h94.768V86.214z"
                  transform="translate(13.56)"
                />
                <path
                  className="ant-empty-img-default-path-3"
                  d="M33.83 0h67.933a4 4 0 0 1 4 4v93.344a4 4 0 0 1-4 4H33.83a4 4 0 0 1-4-4V4a4 4 0 0 1 4-4z"
                />
                <path
                  className="ant-empty-img-default-path-4"
                  d="M42.678 9.953h50.237a2 2 0 0 1 2 2V36.91a2 2 0 0 1-2 2H42.678a2 2 0 0 1-2-2V11.953a2 2 0 0 1 2-2zM42.94 49.767h49.713a2.262 2.262 0 1 1 0 4.524H42.94a2.262 2.262 0 0 1 0-4.524zM42.94 61.53h49.713a2.262 2.262 0 1 1 0 4.525H42.94a2.262 2.262 0 0 1 0-4.525zM121.813 105.032c-.775 3.071-3.497 5.36-6.735 5.36H20.515c-3.238 0-5.96-2.29-6.734-5.36a7.309 7.309 0 0 1-.222-1.79V69.675h26.318c2.907 0 5.25 2.448 5.25 5.42v.04c0 2.971 2.37 5.37 5.277 5.37h34.785c2.907 0 5.277-2.421 5.277-5.393V75.1c0-2.972 2.343-5.426 5.25-5.426h26.318v33.569c0 .617-.077 1.216-.221 1.789z"
                />
              </g>
              <path
                className="ant-empty-img-default-path-5"
                d="M149.121 33.292l-6.83 2.65a1 1 0 0 1-1.317-1.23l1.937-6.207c-2.589-2.944-4.109-6.534-4.109-10.408C138.802 8.102 148.92 0 161.402 0 173.881 0 184 8.102 184 18.097c0 9.995-10.118 18.097-22.599 18.097-4.528 0-8.744-1.066-12.28-2.902z"
              />
              <g
                className="ant-empty-img-default-g"
                transform="translate(149.65 15.383)"
              >
                <ellipse
                  cx="20.654"
                  cy="3.167"
                  rx="2.849"
                  ry="2.815"
                />
                <path
                  d="M5.698 5.63H0L2.898.704zM9.259.704h4.985V5.63H9.259z"
                />
              </g>
            </g>
          </svg>
        </div>
        <p
          className="ant-empty-description"
        >
          当前页面未找到视频:)
        </p>
      </div>
    </section>
  </div>
</div>
`;
